<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <title>About Us</title>
  </head>
  <body>
    {% extends "template.html" %}
    {% block content %}
    <h1> About Our Project </h1>
    <p>The Worm Cool Kit (WCK) is a web interface that offers several online bioinformatic tools for scientists who work
        with the nematode <i>Caenorhabditis elegans</i> (henceforth “worms”).</p>
    <p>The WCK is especially useful for researchers who wish to model human disease-causing mutations in worms, but it
        will be helpful for any analysis of gene and residue conservation between worms and humans and for CRISPR design
        of point mutations. Starting by identifying the most likely worm ortholog of the human gene of interest,
        checking whether a given amino acid variation in the human gene is in a residue conserved in worms and finally
        designing the CRISPR strategy to insert the variation into the worm genome.</p>
    <p>A detailed description of each bioinformatic tool can be found on their respective pages.</p>
    <p><br><u>Privacy Notification:</u> WormCoolKit does not save any of the ortholog searches or CRISPR designs performed on the website.
        Results are only available to the person performing the search.</p>
    <div class="gap-10"></div>
    <p><b>The Orthologs Finder</b></p>
    <p>The Orthologs Finder is a tool designed to deliver the most likely <i>C. elegans</i> genes orthologous to their human
        gene of interest or the other way around - find human orthologs of worm genes.</p>
    <p>Our algorithm extracts candidates from the OrthoList2 database (see References) and carries several additional
        filtration steps, aimed to increase the likelihood that candidates are true orthologs.</p>
    <p>The default thresholds for filtration were determined after examining a large number of established ortholog
        pairs through the pipeline. Some thresholds can be changed by the user.</p>
    <p>Our analysis includes the following steps: </p>
    <ol>
    <li><p>Filtration based on number of sources in OrthoList2: OrthList2 specifies the number of tools (out of six)
        that predicted a pair of genes to be orthologous. Our analysis filters out (as a default) candidates with only 1
        or 2 sources.</p></li>
    <li><p>Filtration based on gene's length: each gene's coding sequence length is extracted from Ensembl, and if the
        ratio between the worm gene to the human gene is lower than 0.5 or higher than 2, the pair is ruled out.</p></li>
    <li><p>Filtration based on number of conserved protein domains: each gene's number of conserved protein domains is
        extracted from GeneBank, and if the ratio between the numbers of domains in the query to the number of domains
        in the candidate is higher than 2 or lower than 0.5, the pair is ruled out.</p></li>
    <li><p>Filtration based on reversed BLAST: In this step we BLAST the candidate ortholog in the query gene’s species,
        and if the query gene is among the first 50 matches, the orthologous pair is valid.</p></li>
    </ol>
    <p>The pipeline will return all pairs that have passed the above-mentioned filters threshold and specify the values
        of each test.</p>

    <div class="gap-10"></div>
    <p><b>The Amino Acid Conservation Tool</b></p>
    <p>This tool is designed to inform whether a specific amino acid in a human protein is conserved in its 
        <i>C. elegans</i> ortholog sequence.</p>
    <p>First, it uses the Ortholog Finder to locate the most probable <i>C. elegans</i> ortholog. Next, it checks for
        conservation of the amino acid of interest. If conservation is confirmed, the tool will provide the user with
        the corresponding site in the worm protein amino acid sequence.
        If not, the tool will notify whether the amino acid in the worm sequence is similar or not conserved at all.</p>
    <p>To identify conservation status, the algorithm uses different methods of sequence alignments between the human
        gene and its ortholog. Therefore, each conserved amino acid is delivered with the number of alignments (out of
        ~2000) supporting the conservation. The tool also delivers an alignment score for the region surrounding the
        relevant amino acid, to assess whether the variation lies within a conserved region of the protein. The score
        shows what percentage of the 60 amino acids around the specified amino acid in the human sequence are aligned
        with the corresponding amino acids in the worm sequence. </p>
    <p>Lastly, the tool also checks if the variant was formerly registered in the Million Mutation Project.</p>

    <div class="gap-10"></div>
    <p><b>Automated CRISPR Planner</b></p>
    <p>The automated CRISPR planner is based on the CRISPR method published by Paix A. et al. The planner is designed to
        first supply the users with RNA guides suitable for their query and then, once a guide is chosen, the algorithm
        defines the relevant parameters such as PAM site, CAS9 double-strand break site and mutation zone. </p>
    <p>It then goes through all possible options to mutate the strand as needed, including mutations to change the
        designated amino acid, mutations that prevent re-attachment of the CAS9 complex and insertion or removal of a
        restriction enzyme site to enable post editing identification of the gene by PCR.</p>
    <p>In each step, the algorithm goes through all possible mutations and chooses the ones that have the highest
        chances to be successfully inserted and would be most convenient for the user’s needs.
        Lastly, the tool provides the user with a complete DNA repair template sequence containing the mutations flanked
        by two homology arms (35nts in length).</p>
    <p>For now, the tool only works for point mutations (one amino acid changed to another) in the 
        <i>Caenorhabditis elegans</i> genome.</p>

    <div class="gap-80"></div>
    <footer>
        <p>&copy; All rights reserved to <a href="https://www.zaidelbarlab.com/">Ronen Zaidel Bar's lab</a>.</p>
        <p>Developed and maintained by Liran Avda of Ronen Zaidel Bar's lab.</p>
    </footer>

    {% endblock %}
  </body>
</html>