<!DOCTYPE html>
<html lang="en" dir="ltr">
  <head>
    <meta charset="utf-8">
    <title>About Us</title>
  </head>
  <body>
    {% extends "template.html" %}
    {% block content %}
    <h1> About Our Project </h1>
    <p>The Worm Cool Kit (WCK) is a web tools interface that offers several web services for scientists that work
        with Caenorhabditis elegans genes. The interface offers three main services:<p>
    <div class="gap-10"></div>
    <p><b>The Ortholog Finder</b></p>
    <p>The Orthologs Finder is a tool designed to deliver you with Human\Caenorhabditis elegans orthologous genes for
        your Caenorhabditis elegans \Human query genes.</p>
    <p>Our algorithm extracts the raw data from the already-analyzed list of orthologous candidates (extracted from
        OrthoList2 data base) and carries several more filtration tests upon those carried by OrthoList2. Those
        tests aim to increase the found orthologs accuracy and filter out candidates that might not be strong
        enough orthologs for the user's purpose.</p>
    <p>Our defaulted bars for filtrations are decided upon after examining large amount of data through the pipeline.
        Some bars can be changed by the user.</p>
    <p>On our analysis, we focused on these extra examinations:</p>
    <p>-	Filtration based on number of sources (OrthoList2): OrthList2 mentions along its orthologs, the number of
        tools (out of six) that predict the genes pair to be orthologous. Our analysis rules out (as a default)
        the candidates for whom the number of sources is 2 or lower.</p>
    <p>-	Filtration based on length ratio: Each gene's number of conserve domains is extracted from the data
        presented in GeneBank, and if the ratio between the numbers of two is higher than 2 or lower than 0.5, the pair
        is ruled out.</p>
    <p>-	Filtration based on conserved domains ration: Each gene's amino acid sequence length is extracted from the
        data presented in GeneBank, and if the ratio between the numbers of two is higher than 2 or lower than 0.5, the
        pair is ruled out.</p>
    <p>-	Filtration based on reversed BLAST: Each pair is consisted of the query gene and the candidate ortholog.
        On this step we BLAST the candidate ortholog for its own orthologs in the relevant species, if the query gene is
        among the first 50 matches, the orthologous pair is valid. </p>
    <p>The pipeline will return all pairs that have passed the bars for all examinations, and mention the values in each
        test.</p>

    <p><b>The Variant Checker</b></p>
    <p>The Variant Checker is a tool designed to inform you if a certain amino acid in a specific place in the
        Caenorhabditis elegans protein's sequence is conserved or not in its human ortholog sequence as well. If this
        amino acid is conserved, the tool will provide you with the corresponding site in the Caenorhabditis elegans
        gene amino acid sequence. </p>
    <p>This tool uses the Ortholog Finder to find the human orthologs on which the conservation is checked.</p>

    <div class="gap-10"></div>
    <p><b>Automated CRISPR Planner - IN PROGRESS</b></p>
    <p>The Automated CRISPR Planner aims to design your CRIPSPR process given only the amino acid mutation you wish to
        insert and the gene name.</p>
    <p>For now, the tool only works for point mutations (one amino acid to another) in the Caenorhabditis elegans genome.
        The algorithm is designed to mimic the human way to insert the fittest nucleotides mutation in the right place
        in order to mutate the gene, to prevent re-attachment and to insert\remove a new restriction site to test the
        specific gene mutation insertion.</p>
    <div class="gap-80"></div>
    <p>Developed by: Liran Avda, for every question you might have, don't hesitate to contact: Liranavda@gmail.com</p>

    {% endblock %}
  </body>
</html>