<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>FAQ</title>
</head>
    {% extends "template.html" %}
    {% block content %}
    <h1> FAQ </h1>
    <ol>
        <li><b>What results does the automated CRISPR planner provide?</b><br>
            Once the crRNA is chosen by the user, the algorithm will try and introduce the relevant strand with codon
            mutations (to alter the amino acid) and mutations to prevent re-cutting by CAS-9 in the relevant section
            (PAM site/crRNA) that will also enable the insertion of a new restriction site.<br>
            If a new restriction site cannot be inserted using the two types of mutations mentioned above, the algorithm
            will try to add point mutations to insert a new restriction site.<br>
            In case of failure, the algorithm will check for the possible removal of an existing restriction site,
            and if not successful, the algorithm will try to add point mutations to remove an existing restriction site.<br>
            If all steps have failed, the tool will provide the user with a message stating that the insertion/removal
            of a restriction site was not possible.<br>
            However, when results are achieved, each result will specify the mutations inserted to change the designated
            amino acid, mutations inserted to prevent re-attachment of the CAS9 complex and the restriction site
            mutations (if necessary), along with the restriction enzyme and the repair template (consists of the section
            with the desired edits and flanked by two 35 nucleotide homology arms). <br>
            The strand (sense or anti-sense) on which the DNA repair template is based on will be stated. Each point
            mutation will be represented by the index of the nucleotide, the former nucleotide and the new
            nucleotide.</li><br>

        <li><b>How does providing preferred restriction enzymes to the Automated CRISPR Planner affect the
            algorithm?</b><br>
            When providing the tool with preferred restriction enzymes, it aims to find CRISPR plans that use
            one of the given enzymes. <br>
            The algorithm will only search for restriction sites using the enzymes you provided it. <br>
            If you wish to search results using all restriction enzymes known to the system, just
            leave the box empty.</li><br>

        <li><b>What does it mean if the query takes too long to return?</b><br>
            In this case, there are probably some difficulties in finding a restriction enzyme to insert.<br>
            If you specified the restriction enzymes you wish to work with, try extending that list or remove
            it altogether to search all restriction enzymes known to the program. <br>
            Also, if you specified a high number of results you wish to get, try lowering that number to check
            whether the algorithm has come up with some results.</li><br>

        <li><b>Where can I find a full list of the restriction enzymes the algorithm searches for?</b><br>
            Our algorithm uses the New England Bio Labs's list of restriction enzymes:
            <a href="https://international.neb.com/tools-and-resources/selection-charts/alphabetized-list-of-recognition-specificities">New England BioLabs.</a><br>
            If you wish to add a restriction enzyme that is not included in this list, please contact us.</li><br>

        <li><b>What would happen if I specified a restriction enzyme's name that is unknown to the algorithm?</b><br>
            The algorithm would disregard it. <br>
            Your results show which of the enzymes you entered are valid and were used
            in the search.</li><br>

        <li><b>In what format should I enter the restriction enzymes I wish to use?</b><br>
            The algorithm will accept a list of names of the restriction enzymes you are interested in, without the
            HF notation for high fidelity. <br>
            The enzymes' names can be separated by commas, tabs, semi colons or spaces.<br>
            To check whether your restriction enzymes' names are correctly formatted, check your results.</li><br>

        <li><b>Could there be other CRISPR designs that the algorithm didn't find?</b><br>
            Yes. The algorithm's purpose is not to find all possible designs for the point mutation you are
            interested in, but to find the best options relevant for your query and return them when the requested
            amount of possible results is achieved (default of one result).<br>
            If you are not satisfied with the results given, you can alter the maximum results box to be given more
            results.<br>
            However, if you are given less results than you specified in the maximum results box, then the
            algorithm could not find more results, thus changing the bar of maximum results won't make any
            difference. <br>
            If you feel something is wrong with the results, please contact us with the details of
            your query.</li><br>

        <li><b>Why is providing the nucleotide sequence optional? When is it necessary to provide it?</b><br>
            The algorithm will try to extract the nucleotide sequence (unspliced + UTR) from several sources.<br>
            However, in case the algorithm cannot find the relevant sequence (or is faulty and doesn't match the
            protein sequence it later extracts), an error would be presented to the user, specifying that the
            system's findings do not match the amino acid you wish to alter. <br>
            In that case, you have the possibility to provide the correct sequence yourself.</li><br>

        <li><b>Are the restriction sites found filtered? How are they filtered?</b><br>
            Yes. The filtration steps are as following:<br>
            1. The restriction sites found in the process of inserting/removing restriction sites are filtered out if a
            same site is found in a distance of less than 150 nt from the found site. <br>
            2. Restriction sites are also sorted by their proximity to the Double Strand Break (DSB) point
            (closer is better). <br>
            3. The algorithm takes into account the distance from other same sites in the range of 500 nt around the
            found site and by the frequency of that site in the range of 500 nt around the found site.</li><br>

        <li><b>If the codon we wish to mutate has point mutations before and after the DSB, which arm would be
            the one we mutate?</b><br>
            If the number of point mutations on one side of the DSB is higher than the number of point mutations on
            another, then the side with the higher number of mutations will be chosen as the section to be mutated.<br>
            If not, the algorithm first defines the sequence with the PAM site to be the mutated sequence, and if
            mutating the PAM sites is impossible, the algorithm will take the second sequence to design the CRISPR
            plan onto.</li><br>

        <li><b>How far away from the DSB point would the algorithm try to insert mutation
            (or: what range is taken to be the mutation zone)?</b><br>
            The mutation zone is defined to be ranging no more than 20 nucleotides away from the DSB. Which means that
            all the mutation introduced to the strand, namely mutations that change the amino acid, mutations to prevent
            re-cutting by CAS-9 complex (to the PAM site or recognition site) and mutations for introducing or removing
            a restriction site, will be within 20 nucleotides of the DSB site.</li><br>

        <li><b>What is the numbering scheme the indices are represented by?</b><br>
            The results incorporate zero-based numbering; the initial element of a sequence is assigned the index
            0.</li><br>

        <li><b>What protocol does the automated CRISPR planner follow?</b><br>
            Our algorithm is based on steps and the scheme described in this paper:<br>
            Paix A, Folkmann A, Seydoux G. Precision genome editing using CRISPR-Cas9 and linear repair templates
            in C. elegans. Methods. 2017;121-122:86-93. doi:10.1016/j.ymeth.2017.03.023<br>
            Our lab has made a few minor alterations to the CRISPR protocol, you can view our revised protocol here:<br>
            <a href="{{ url_for('protocol') }}">CRISPR protocol alterations and tips</a></li><br>

        <li><b>What does the length ratio range in the Orthologs Finder service refer to?</b><br>
            The length ratio range is a tool with which you can decide the length ratio parameters upon which the
            results are filtered. By default, the service filters out orthologs pairs whose coding sequences ratio is
            lower than 0.5 or higher than 2 (the ratio is calculated to be the length of the coding sequence of the gene
            the user provided divided by the length of the coding sequence of the putative ortholog.</li><br>

        <li><b>What does the minimum number of sources in the Orthologs Finder service refer to?</b><br>
            The minimum number of sources is a tool with which you can decide the sources parameter upon which the
            results are filtered. By default, the service filters out orthologs pairs for whose the number of programs
            (out of 6) that predict them to be orthologous is lower than 2 (the number of programs is derived
            from <a href="http://ortholist.shaye-lab.org/" target="_blank">OrthoList2</a>).</li><br>

        <li><b>What information can I get from Amino Acid Conservation tool?</b><br>
            Check The Variants tool aims to provide the user with information regarding whether a specific amino acid in
            the human sequence is conserved in its C.elegans ortholog. If the amino acid is conserved, the information
            will include the corresponding site in the relevant C.elegans ortholog, the conservation score of a 60-aa
            window around the conserved amino acid, the number of alignments (out of 2000) supporting of the conservation
            and indication of whether the variant was listed in the
            <a href="http://genome.sfu.ca/mmp/" target="_blank">Million Mutations Project</a>.</li><br>
    </ol>

    <footer>
        <p>&copy; All rights reserved to <a href="https://www.zaidelbarlab.com/">Ronen Zaidel Bar's lab</a>.</p>
        <p>Developed and maintained by Liran Avda of Ronen Zaidel Bar's lab.</p>
    </footer>
    {% endblock %}
  </body>
</html>